Laboratory/C++/gen/jazz.hpp

```diff
@@ -0,0 +1,156 @@
+    auto &writer = bebop::BebopWriter::instance();
+    Song::encodeInto(message, writer);
+    return std::move(writer.buffer());
+  }
+
+  static void encodeInto(const Song &message, bebop::BebopWriter &view) {
+    const auto pos = view.reserveMessageLength();
+    const auto start = view.length();
+    if (message.title.has_value()) {
+      view.writeByte(1);
+      view.writeString(message.title.value());
+    }
+    if (message.year.has_value()) {
+      view.writeByte(2);
+      view.writeUint16(message.year.value());
+    }
+    if (message.performers.has_value()) {
+      view.writeByte(3);
+      {
+      const auto length0 = message.performers.value().size();
+      view.writeUint32(length0);
+      for (const auto &i0 : message.performers.value()) {
+        Musician::encodeInto(i0, view);
+      }
+    }
+    }
+    view.writeByte(0);
+    const auto end = view.length();
+    view.fillMessageLength(pos, end - start);
+  }
+
+  static Song decode(const uint8_t *buffer) {
+    Song result;
+    Song::readIntoFrom(result, bebop::BebopReader::instance(buffer));
+    return result;
+  }
+
+  static void readIntoFrom(Song &target, bebop::BebopReader& view) {
+    const auto length = view.readMessageLength();
+    const auto end = view.pointer() + length;
+    while (true) {
+      switch (view.readByte()) {
+        case 0:
+          return;
+        case 1:
+          target.title = view.readString();
+          break;
+        case 2:
+          target.year = view.readUint16();
+          break;
+        case 3:
+          {
+        const auto length0 = view.readUint32();
+        target.performers = std::vector<Musician>();
+        target.performers->reserve(length0);
+        for (size_t i0 = 0; i0 < length0; i0++) {
+          Musician x0;
+          Musician::readIntoFrom(x0, view);
+          target.performers->push_back(x0);
+        }
+      }
+          break;
+        default:
+          view.seek(end);
+          return;
+      }
+    }
+  }
+};
+
+struct Library {
+  std::map<bebop::Guid, Song> songs;
+
+  static std::unique_ptr<std::vector<uint8_t>> encode(const Library &message) {
+    auto &writer = bebop::BebopWriter::instance();
+    Library::encodeInto(message, writer);
+    return std::move(writer.buffer());
+  }
+
+  static void encodeInto(const Library &message, bebop::BebopWriter &view) {
+    view.writeUint32(message.songs.size());
+    for (const auto &e0 : message.songs) {
+      view.writeGuid(e0.first);
+      Song::encodeInto(e0.second, view);
+    }
+  }
+
+  static Library decode(const uint8_t *buffer) {
+    Library result;
+    Library::readIntoFrom(result, bebop::BebopReader::instance(buffer));
+    return result;
+  }
+
+  static void readIntoFrom(Library &target, bebop::BebopReader& view) {
+    {
+      const auto length0 = view.readUint32();
+      target.songs = std::map<bebop::Guid, Song>();
+      for (size_t i0 = 0; i0 < length0; i0++) {
+        bebop::Guid k0;
+        Song v0;
```

Comment:
So I know C++ has copy elision, but the rules for it are complex enough that I can never remember them, so if I can avoid it I always try to write code that I definitely know won't incur a copy if I can. So I think if you had this code:
```C++
bebop::Guid k0 = view.readGuid();
Song& v0 = target.songs[k0];
Song::readIntoFrom(v0, view);
```
Then you would be writing directly to the instance of the Song that is stored in the map. Since the `Song& v0 = target.songs[k0];` will insert a default constructed Song and then return a reference to it.

Also, depending on the complexity of the default constructor for bebop::Guid, it might be slightly more efficient to construct it on the same line that you are doing view.readGuid(). Note that:
```C++
bebop::Guid k0; // invokes default constructor
...
k0 = view.readGuid(); // invokes the Guid assignment operator
```
but
```C++
bebop::Guid k0 = view.readGuid(); // invokes only the copy constructor
```